packages/@react-aria/landmark/src/useLandmark.ts

```diff
@@ -0,0 +1,293 @@
+   * Returns undefined if there are no landmarks.
+   */
+  public getNextLandmark(element: HTMLElement, {backward}: {backward?: boolean }) {
+    if (this.landmarks.length === 0) {
+      return undefined;
+    }
+
+    let currentLandmark = this.closestLandmark(element);
+    let nextLandmarkIndex = backward ? -1 : 0;
+    if (currentLandmark) {
+      nextLandmarkIndex = this.landmarks.findIndex(landmark => landmark === currentLandmark) + (backward ? -1 : 1);
+    }
+
+    // Wrap if necessary
+    if (nextLandmarkIndex < 0) {
+      nextLandmarkIndex = this.landmarks.length - 1;
+    } else if (nextLandmarkIndex >= this.landmarks.length) {
+      nextLandmarkIndex = 0;
+    }
+
+    return this.landmarks[nextLandmarkIndex];
+  }
+
+  /**
+   * Look at next landmark. If an element was previously focused inside, restore focus there.
+   * If not, focus the first focusable element inside the lanemark.
+   * If no focusable elements inside, go to the next landmark.
+   * If no landmarks at all, or none with focusable elements, don't move focus.
+   */
+  public f6Handler(e: KeyboardEvent) {
+    if (e.key === 'F6') {
+      e.preventDefault();
+      e.stopPropagation();
+
+      let backward = e.shiftKey;
+      let nextLandmark = this.getNextLandmark(e.target as HTMLElement, {backward});
+
+      // If no landmarks, return
+      if (!nextLandmark) {
+        return;
+      }
+
+      // If alt key pressed, focus main landmark
+      if (e.altKey) {
+        let main = this.getLandmarkByRole('main');
+        if (main && document.contains(main.ref.current)) {
+          this.focusLandmark(main.ref.current);
+        }
+        return;
+      }
+
+      // If something was previously focused in the next landmark, then return focus to it
+      if (nextLandmark.lastFocused) {
+        let lastFocused = nextLandmark.lastFocused;
+        if (document.body.contains(lastFocused)) {
+          lastFocused.focus();
+          return;
+        }
+      }
+
+      // Otherwise, focus the landmark itself
+      if (document.contains(nextLandmark.ref.current)) {
+        this.focusLandmark(nextLandmark.ref.current);
+      }
+    }
+  }
+
+  /**
+   * Sets lastFocused for a landmark, if focus is moved within that landmark.
+   * Lets the last focused landmark know it was blurred if something else is focused.
+   */
+  public focusinHandler(e: FocusEvent) {
+    let currentLandmark = this.closestLandmark(e.target as HTMLElement);
+    if (currentLandmark && currentLandmark.ref.current !== e.target) {
+      this.updateLandmark({...currentLandmark, lastFocused: e.target as HTMLElement});
+    }
+    let previousFocusedElment = e.relatedTarget as HTMLElement;
+    if (previousFocusedElment) {
+      let closestPreviousLandmark = this.closestLandmark(previousFocusedElment);
+      if (closestPreviousLandmark && closestPreviousLandmark.ref.current === previousFocusedElment) {
+        closestPreviousLandmark.onBlur();
+      }
+    }
```

Comment:
Found a scenario where the previously landmark doesn't seem to update its `isLandmarkFocused` state and retains its `tabIndex` of -1.
Repro steps:
1. Go to https://reactspectrum.blob.core.windows.net/reactspectrum/5ed0f5712593a0564200aa90a867b3894d181d68/storybook/index.html?path=/story/landmark--flat-landmarks. Open devtools so you can see the tab index applied on each landmark
2. Tab though the links and the textfield
3. Hit alt+f6 to focus the main landmark
4. Click on the empty space of the page. Note that the tabIndex hasn't been removed from the main landmark
5. Hit f6 a bunch. The tabindex doesn't get removed from the main landmark still

Reply:
Actually an easier reproduction case:
1. In the same story hit alt+f6
2. Click on the empty space of the page
3. Hit f6 a bunch of times. Note you get stuck on the nav landmark

Maybe we could just iterate over all landmarks and call onBlur on every single one that isn't the currentLandmark?

Reply:
sounds like a good unit test too :)

Reply:
I think maybe you need a blur/focusout handler. Another case is if you click outside the window entirely, e.g. on the address bar.